core/lib/state/src/storage_factory.rs

```diff
@@ -153,6 +164,90 @@ impl PgOrRocksdbStorage<'static> {
         tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
         Ok(Some(rocksdb.into()))
     }
+
+    /// Creates a storage snapshot. Require protective reads to be persisted for the batch, otherwise
+    /// will return `Ok(None)`.
+    #[tracing::instrument(skip(connection))]
+    pub async fn snapshot(
+        connection: &mut Connection<'static, Core>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<StorageSnapshot>> {
+        let Some(header) = connection
+            .blocks_dal()
+            .get_l1_batch_header(l1_batch_number)
+            .await?
+        else {
+            return Ok(None);
+        };
+        let bytecode_hashes: HashSet<_> = header
+            .used_contract_hashes
+            .into_iter()
+            .map(u256_to_h256)
+            .collect();
+
+        // Check protective reads early on.
+        let protective_reads = connection
+            .storage_logs_dedup_dal()
+            .get_protective_reads_for_l1_batch(l1_batch_number)
+            .await?;
+        if protective_reads.is_empty() {
+            tracing::debug!("No protective reads for batch");
+            return Ok(None);
+        }
+        let protective_reads_len = protective_reads.len();
+        tracing::debug!("Loaded {protective_reads_len} protective reads");
+
+        let touched_slots = connection
+            .storage_logs_dal()
+            .get_touched_slots_for_l1_batch(l1_batch_number)
+            .await?;
+        tracing::debug!("Loaded {} touched keys", touched_slots.len());
+
+        let all_accessed_keys: Vec<_> = protective_reads
+            .into_iter()
+            .map(|key| key.hashed_key())
+            .chain(touched_slots.into_keys())
+            .collect();
```

Comment:
It probably won't work for every case. E.g. if within an l2 block there were 2 storage write logs: first writes to some slot, second is equal to first but with `rollback: true`; then this slot won't be in `protective_reads` (because of `protective_reads` deduplication rules) nor `touched_slots`(because of l2 block deduplication).
I may be mistaken, please validate

Reply:
Nice catch! There are indeed slots that are not covered in this (or similar) way; about 5–10 per batch on the mainnet (~30k–35k storage accesses). I've reworked snapshot logic correspondingly to require a fallback; PTAL.